src/routes/api/ticket/assign/+server.ts

```diff
@@ -0,0 +1,129 @@
+import {
+    AssignAgentToTicketResult,
+    assignAgentToTicket,
+    canAssignSelfToTicket,
+    getUserFromSession,
+    isAssignedAgent,
+    isAssignedDepartment,
+    isHeadSession,
+    removeTicketAgent,
+} from '$lib/server/database';
+import { AssertionError } from 'node:assert/strict';
+import type { RequestHandler } from './$types';
+import { StatusCodes } from 'http-status-codes';
+import { error } from '@sveltejs/kit';
+
+function resultToCode(result: AssignAgentToTicketResult) {
+    switch (result) {
+        case AssignAgentToTicketResult.Success:
+            return StatusCodes.CREATED;
+        case AssignAgentToTicketResult.NoTicket:
+        case AssignAgentToTicketResult.NoAgent:
+            return StatusCodes.NOT_FOUND;
+        default:
+            throw new AssertionError();
+    }
+}
+
+// eslint-disable-next-line func-style
+export const POST: RequestHandler = async ({ cookies, request }) => {
+    const form = await request.formData();
+
+    const tid = form.get('ticket');
+    if (tid === null || tid instanceof File) throw error(StatusCodes.BAD_REQUEST);
+
+    const dept = form.get('dept');
+    if (dept === null || dept instanceof File) throw error(StatusCodes.BAD_REQUEST);
+    const did = parseInt(dept, 10) >> 0;
+
+    const uid = form.get('uid');
+    if (uid === null || uid instanceof File) throw error(StatusCodes.BAD_REQUEST);
+
+    const sid = cookies.get('sid');
+    if (!sid) throw error(StatusCodes.UNAUTHORIZED);
+
+    // Permissions check
+    // Department assignment check
+    // TODO: Null check for isAssignedAgent and isAssignedDepartment
+    const deptStatus = await isAssignedDepartment(tid, did);
+    if (!deptStatus) throw error(StatusCodes.FORBIDDEN);
+
+    // Department head, assigned agent and self-assignment check
+    const head = await isHeadSession(sid, did);
+    if (head === null) throw error(StatusCodes.UNAUTHORIZED);
+
+    const user = await getUserFromSession(sid);
+    if (user === null) throw error(StatusCodes.UNAUTHORIZED);
+    const assigned = await isAssignedAgent(tid, user.user_id);
```

Comment:
You may want to consider writing a dedicated database function that checks for the head session or the assigned agent. By invoking `isHeadSession` and `isAssignedAgent` separately, we actually incur two network roundtrip times to the database, which is not ideal.

```sql
-- It's probably gonna look something like this.
SELECT is_head_session(...) OR is_assigned_agent(...);
```